open_spiel/algorithms/infostate_tree.cc

```diff
@@ -0,0 +1,814 @@
+
+  // 2. Prune away unreachable subtrees.
+  //
+  // This can be done with a more costly recursion.
+  // Instead we make a more cache-friendly double pass through the response
+  // vector: we increment the visited path by 1, resulting in a value of 2.
+  // Then we zero-out all values but 2.
+  current = empty_sequence();
+  response[current] = 2.;
+  while (!IsLeafSequence(current)) {
+    for (SequenceId seq : observation_infostate(current)->AllSequenceIds()) {
+      if (response[seq] == 1.) {
+        current = seq;
+        response[seq] += 1.;
+        break;
+      }
+    }
+  }
+  for (SequenceId seq : response.range()) {
+    response[seq] = response[seq] == 2. ? 1. : 0.;
+  }
+  SPIEL_DCHECK_TRUE(IsValidSfStrategy(response));
+  return {gradient[empty_sequence()], response};
+}
+
+double InfostateTree::BestResponseValue(LeafVector<double>&& gradient) const {
+  // Loop over all heights.
+  for (int d = tree_height_ - 1; d >= 0; d--) {
+    int left_offset = 0;
+    // Loop over all parents of current nodes.
+    for (int parent_idx = 0; parent_idx < nodes_at_depths_[d].size();
+         parent_idx++) {
+      const InfostateNode* node = nodes_at_depths_[d][parent_idx];
+      const int num_children = node->num_children();
+      const Range<LeafId> children_range = gradient.range(
+          left_offset, left_offset + num_children);
+      const LeafId parent_id(parent_idx, this);
+
+      if (node->type() == kDecisionInfostateNode) {
+        double max_value = std::numeric_limits<double>::min();
+        for (LeafId id : children_range) {
+          max_value = std::fmax(max_value, gradient[id]);
+        }
+        gradient[parent_id] = max_value;
+      } else {
+        SPIEL_DCHECK_EQ(node->type(), kObservationInfostateNode);
+        double sum_value = 0.;
+        for (LeafId id : children_range) {
+          sum_value += gradient[id];
+        }
+        gradient[parent_id] = sum_value;
+      }
+      left_offset += num_children;
+    }
+    // Check that we passed over all of the children.
+    SPIEL_DCHECK_EQ(left_offset, nodes_at_depths_[d + 1].size());
+  }
+  const LeafId root_id(0, this);
+  return gradient[root_id];
+}
+
+DecisionId InfostateTree::DecisionIdFromInfostateString(
+    const std::string& infostate_string) const {
+  for (InfostateNode* node : decision_infostates_) {
+    if (node->infostate_string() == infostate_string)
+      return node->decision_id();
+  }
+  return kUndefinedDecisionId;
+}
+
+bool CheckSum(const SfStrategy& strategy, SequenceId id, double expected_sum) {
+  if (fabs(strategy[id] - expected_sum) > 1e-13) {
+    return false;
+  }
+
+  const InfostateTree* tree = strategy.tree();
+  if (tree->IsLeafSequence(id)) {
+    return true;
+  }
+
+  double actual_sum = 0.;
+  const InfostateNode* node = tree->observation_infostate(id);
+  for (SequenceId sub_seq : node->AllSequenceIds()) {
+    actual_sum += strategy[sub_seq];
+  }
+  if (fabs(actual_sum - expected_sum) > 1e-13) {
```

Comment:
This repeating pattern could be moved into a utility function `if (deviation(first, second) < 1e-13) \\\...`